### PR TITLE
Support OSX

### LINKED ISSUE #2892: CI - Add running CI tests on Mac
* Travis supports testing on Mac: https://docs.travis-ci.com/user/osx-ci-environment/

* Only two tests don't pass: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/2846

* These will probably be fixed upstream: https://github.com/sebastianbergmann/phpunit/issues/2697

* No Mac support is 😢

### PR DESCRIPTION
**Pull request by aidantwoods**
Add support for OSX by testing it in Travis CI (merge probably depends on an update to phpunit, see: https://github.com/sebastianbergmann/phpunit/issues/2697)

If anyone can see a nicer way of doing this let me know – testing various configs in travis isn't my favourite activity ;p

---

Closes #2892

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -52,12 +53,43 @@ jobs:
...
+            os: osx
+            language: generic
+            before_install:
+                # Set up OSX environment
+                - travis_retry brew update
+                - travis_retry brew tap homebrew/homebrew-php
+                - travis_retry brew install $phpver || brew outdated $phpver || brew upgrade $phpver
+                - brew link $phpver
+                - travis_retry wget https://raw.githubusercontent.com/composer/getcomposer.org/master/web/installer -O - -q | php -- --install-dir=/usr/local/bin && mv /usr/local/bin/composer.phar /usr/local/bin/composer
+                - travis_retry composer self-update
```
Comment (GrahamCampbell) [inline .travis.yml]: Why is this needed?

**Comment by keradus** [comment]
Supported env is not a feature. header adjusted.

File: .travis.yml [hunk 2]
```diff
@@ -52,12 +53,43 @@ jobs:
...
+
+                # turn off XDebug
+                - phpenv config-rm xdebug.ini || return 0
+
+                # Composer: boost installation
+                - composer global show -ND 2>&1 | grep "hirak/prestissimo" || travis_retry composer global require $DEFAULT_COMPOSER_FLAGS hirak/prestissimo
```
Comment (GrahamCampbell) [inline .travis.yml]: Is this done for the linux builds?

Comment (SpacePossum) [inline .travis.yml]: it is, check the rest of the file ;)

Comment (GrahamCampbell) [inline .travis.yml]: 👍

File: .travis.yml (hunk 1 above)
Comment (aidantwoods) [inline .travis.yml]: Travis don't include anything useful for PHP dev in their OSX builds (in-fact specifying `language: php` flat out breaks things, so idk) – let me know if there's a better solution though!

**Comment by keradus** [comment]
big :-1: for extending matrix that much, we don't need 3 new jobs just to test osx for sure.

File: .travis.yml (hunk 1 above)
Comment (GrahamCampbell) [inline .travis.yml]: I am specifically referring to the self-update. You already installed the latest version.

**Comment by SpacePossum** [comment]
Looks promising :)
One the two expected tests to fail, an explicit test on `PHP_OS` or `PHP_OS_FAMILY` and skip when needed would be OK IMHO. It's not ideal, but it only effects two tests so worth it to get this going.

**Comment by aidantwoods** [comment]
> big 👎 for extending matrix that much, we don't need 3 new jobs just to test osx for sure.

How many tests would you like? Just tried to mirror the linux tests so far

**Comment by keradus** [comment]
it increase workload a lot for Travis...
what about taking other apporach?
current state is Travis for Linux, AppVeyor for for Windows... maybe CircleCI for OSX ?

**Comment by GrahamCampbell** [comment]
Maybe circle for everything tbh. Travis is very slow compared to circle.

**Comment by aidantwoods** [comment]
> it increase workload a lot for Travis...
> what about taking other apporach?
> current state is Travis for Linux, AppVeyor for for Windows... maybe CircleCI for OSX ?

Perhaps, yes. A separate build stage might be another option too?

RE the extra workload, shouldn't be too noticeable with concurrent builds (I hope). I think Travis may be having a slow period – up to 20mins on this linux job... https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/248866481

OSX hasn't even started yet.

**Comment by keradus** [comment]
separate stage would be executed after current testing stage, making all executes longer.
concurrency works inside same stage, not across different stages.

laso, consurrency works up to 4 jobs for free plans

**Comment by julienfalque** [comment]
I would say the current slowliness is more a performance issue in the codebase than Travis being slow. This is a pure guess and has to be verified though.

Anyway, I think it would be better to target 2.2.

File: .travis.yml (hunk 1 above)
Comment (aidantwoods) [inline .travis.yml]: Ah, yes you're right there – I'll remove that!

**Comment by keradus** [comment]
the failing of utests is probably on higher branch, is that right @aidantwoods  ?

**Comment by aidantwoods** [comment]
> separate stage would be executed after current testing stage, making all executes longer.
concurrency works inside same stage, not across different stages.

Indeed, so we'd be looking at an extra 2-3mins if all the OSX builds start together on a separate stage (though compared to 20mins for that `php56` build, not sure how significant that is?)

**Comment by aidantwoods** [comment]
> the failing of utests is probably on higher branch, is that right @aidantwoods ?

The `ProcessLinterTest` failures are present in `2.2` too

**Comment by aidantwoods** [comment]
https://github.com/aidantwoods/PHP-CS-Fixer/commit/287a853b52a7639af621fa67ec752bb0f4808b99 adds two OSX builds onto `2.2`'s Travis, incase we wanted to rebase this onto `2.2`.

**Comment by keradus** [comment]
@aidantwoods, could you execute tests of 2.2 branch on mac and send a fix as a separate PR please?

**Comment by keradus** [comment]
replaced by #2932
